v2/APIError_test.go

```diff
@@ -0,0 +1,281 @@
+package gax
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func TestEmptyDetails(t *testing.T) {
+	errdetails := ErrDetails{}
+	apierr := APIError{
+		err:     nil,
+		status:  nil,
+		details: errdetails,
+	}
+	var actualDetails = apierr.Details()
+	assert.Equal(t, apierr.details, actualDetails)
+}
+
+func TestNonEmptyDetails(t *testing.T) {
+	qf := &errdetails.QuotaFailure{}
+	v := &errdetails.QuotaFailure_Violation{
+		Subject:     "Quotafailure",
+		Description: "You have exhuasted the quota",
+	}
+	qf.Violations = append(qf.Violations, v)
+	msg := ErrDetails{}
+	st, _ := status.New(codes.ResourceExhausted, "Per user quota has been exhausted").WithDetails(qf)
+	err := st.Err()
+	msg.QuotaFailure = qf
+	d := APIError{
+		err:     err,
+		status:  st,
+		details: msg,
+	}
+	assert.Equal(t, d.details, d.Details())
+}
+
+func TestUnwrap(t *testing.T) {
+	br := &errdetails.BadRequest{}
+	desc := "The argument is invalid"
+	v := &errdetails.BadRequest_FieldViolation{
+		Field:       "username",
+		Description: desc,
+	}
+	br.FieldViolations = append(br.FieldViolations, v)
+	st, _ := status.New(codes.InvalidArgument, "Invalid argument").WithDetails(br)
+	err := st.Err()
+	msg := ErrDetails{}
+	msg.BadRequest = br
+	d := APIError{
+		err:     err,
+		status:  st,
+		details: msg,
+	}
+	assert.Equal(t, d.err, d.Unwrap())
+}
+func TestError(t *testing.T) {
+	pf := &errdetails.PreconditionFailure{}
+	viol := &errdetails.PreconditionFailure_Violation{
+		Type:    "Precondition Failure",
+		Subject: "This is a test",
+	}
+	pf.Violations = append(pf.Violations, viol)
+	mesg := ErrDetails{}
+	mesg.PreconditionFailure = pf
+	stat, _ := status.New(codes.FailedPrecondition, "System's state is not suitable for operation execution").WithDetails(pf)
+	er := stat.Err()
+	d := APIError{
+		err:     er,
+		status:  stat,
+		details: mesg,
+	}
+	//str := err.Error() + "\n" + br.String()
+	assert.Contains(t, d.Error(), "System's state is not suitable for operation execution")
+}
+
+func TestGRPCStatus(t *testing.T) {
+	qf := &errdetails.QuotaFailure{}
+	v := &errdetails.QuotaFailure_Violation{
+		Subject:     "Quotafailure",
+		Description: "You have exhuasted the quota",
+	}
+	qf.Violations = append(qf.Violations, v)
+	st, _ := status.New(codes.ResourceExhausted, "Per user quota has been exhausted").WithDetails(qf)
+	err := st.Err()
+	msg := ErrDetails{}
+	msg.QuotaFailure = qf
+	d := APIError{
+		err:     err,
+		status:  st,
+		details: msg,
+	}
+	assert.Equal(t, d.status, d.GRPCStatus())
+
+}
+func TestNilFromError(t *testing.T) {
+	err := error(nil)
```

Comment:
`nil` should be an appropriate value to assign to `err` no need for this `error`